Kernel/linker.ld

```diff
@@ -11,7 +11,7 @@ SECTIONS
         $<TARGET_OBJECTS:boot>
         *(.multiboot)
         start_of_kernel_text = .;
-        *(.text)
+        *(.text*)
         *(.text.startup)
```

Comment:
Not an LD expert, but it seems like `*.text*` would suck in `.text.safemem` and `.text.safemem.atomic`? Or does this only apply to sections not explicitly mentioned?

Reply:
Yep, you're right. Moving `*(.text*)` after the safemem sections seems to fix this though, although I'm not sure if having the safemem stuff before the main `.text` is bad? I've also removed `*(.text.startup)` since `*(.text*)` includes that.

Reply:
Does it work by coincidence, or is it documented to work like this?

As for the safemem, moving that code into another section would affect the `start_of_safemem_atomic_text`/`end_of_safemem_atomic_text`, which are used by the page fault handler for a check whether a page fault needs to be handled in a special way. We probably should add a few assertions during startup checking that these functions reside within the range of these labels so that it doesn't get broken accidentally. I think the other changes are fine, but I didn't take a super close look.

Reply:
I believe it's documented to work like this. [Here](https://doc.ecoscentric.com/gnutools/doc/ld/Input-Section-Wildcards.html#Input-Section-Wildcards) it says
> If a file name matches more than one wildcard pattern, or if a file name appears explicitly and is also matched by a wildcard pattern, the linker will use the first match in the linker script.

In our case, the file name is always `*`, however `*(.text.safemem)` and `*(.text.safemem.atomic)` appear before `*(.text*)` so I think they are matched first. I think the assertions are a good idea, we could also assert that `&start_of_safemem_atomic_text != &end_of_safemem_atomic_text` which is what was happening before I moved `*(.text*)` after the safemem sections.

Reply:
Where would the best place to put the assertions be? Somewhere in `init.cpp` maybe?

Reply:
Probably, maybe after this line?
https://github.com/SerenityOS/serenity/blob/ef9bfbd38343cf9bfa708c0d714bd8556031deae/Kernel/init.cpp#L133

Reply:
Ok, I've added the assertions. I've also just wrapped the safemem sections with `KEEP()` to make sure the linker doesn't try to remove them.